src/Symfony/Component/Messenger/Middleware/StackMiddleware.php

```diff
@@ -20,31 +20,74 @@
  */
 class StackMiddleware implements MiddlewareInterface, StackInterface
 {
-    private $middlewareIterator;
+    private $stack;
+    private $offset = 0;
 
-    public function __construct(\Iterator $middlewareIterator = null)
+    /**
+     * @param iterable|MiddlewareInterface[]|MiddlewareInterface|null $middlewareIterator
+     */
+    public function __construct($middlewareIterator = null)
     {
-        $this->middlewareIterator = $middlewareIterator;
+        $this->stack = new MiddlewareStack();
+
+        if (null === $middlewareIterator) {
+            return;
+        }
+
+        if ($middlewareIterator instanceof \Iterator) {
+            $this->stack->iterator = $middlewareIterator;
+        } elseif ($middlewareIterator instanceof MiddlewareInterface) {
+            $this->stack->stack[] = $middlewareIterator;
+        } elseif (!\is_iterable($middlewareIterator)) {
+            throw new \TypeError(sprintf('Argument 1 passed to %s() must be iterable of %s, %s given.', __METHOD__, MiddlewareInterface::class, \is_object($middlewareIterator) ? \get_class($middlewareIterator) : \gettype($middlewareIterator)));
+        } else {
+            $this->stack->iterator = (function () use ($middlewareIterator) {
+                yield from $middlewareIterator;
+            })();
+        }
     }
 
     public function next(): MiddlewareInterface
     {
-        if (null === $iterator = $this->middlewareIterator) {
+        if (null === $next = $this->stack->next($this->offset)) {
             return $this;
         }
-        $iterator->next();
 
-        if (!$iterator->valid()) {
-            $this->middlewareIterator = null;
+        ++$this->offset;
 
-            return $this;
-        }
-
-        return $iterator->current();
+        return $next;
     }
 
     public function handle(Envelope $envelope, StackInterface $stack): Envelope
     {
         return $envelope;
     }
 }
+
+/**
+ * @internal
+ */
+class MiddlewareStack
```

Comment:
Can't you simply drop the iterator, store middleware as array in `StackMiddleware` and advance the pointer manually?

Reply:
That would defeat laziness, which is a desired property.